# Comparative analysis of methodologies for detecting eccDNA
![](figure/Overview.png "Overview")

Implementation description
We collected 7 eccDNA datasets and designed a pipeline to simulate eccDNA data generated by eccDNA-enriched experimental methods.

## Getting Stared
* [Requirements](#requirements)
* [Installation](#Installation)

## Latest updates
### version 1.0.1 May 2024
- Fix a bug in class `seqsim` that can not simulate data with only simple or chimeric eccDNA template.
- Package the pipeline using setuptools
### version 1.0.0 Nov 2023
- Upload ecsim for simulated eccDNA datasets

## Requirements
##### Note: To run this pipeline to simulate data, please ensure that `pysam`, `art`, `seqkit`, and `pbsim2` have already been installed.
##### Create a environment to run ecsim. It would take several minutes to create the environment. 
```
conda env create -f environment.yaml
```
or
```
conda create -n ecsim -c bioconda pysam pbsim2 art seqkit python=3.9
```
## Installation
```
git clone https://github.com/QuKunLab/eccDNABenchmarking.git

```
## Tutorial
##### Note: Please ensure that you have fasta files of reference genome and at least one profile bed files (either simple or chimeric) (containg 5 columns including 'chromatin start end length ecID' ), you can find example profile bed files in [Resource Directory](ecsim/ecsim/resource/).It would take hours to generate simulated dataset with 10000 circular and 10000 linear DNA, with a 5000bp extention of RCA and a sequencing depth of 50.  
```
cd $ecsim_directory
python ecsim.py --sample SAMPLE_NAME (required) \
                --reference REFERENCE_PATH (required) \
                --thread THREAD_NUMBER (required) \
                --path DATA_PATH (required) \
                --simple-profile SIMPLE_PROFILE_BEDFILE (containing 5 columns including `chromatin start end length ecID`) \
                --chimeric-profile CHIMERIC_PROFILE_BEDFILE (containing 5 columns including `chromatin start end length ecID`) \
                --simple-ratio SIMPLE_DNA_RATIO (default is calculated from profile) \
                --meancov SEQUENCING_DEPTH (default=25) \
                --circular-number CIRCULAR_DNA_NUMBER (default=10000) \
                --linear-number LINEAR_DNA_NUMBER (default=10000) \
                --amp AMPLIFIED_LENGTH (default=5000) \
                --seed RANDOM_SEED (default=None) \
                --ont-model PBSIM2_MODEL_NAME (default='R94') \
                --ont-mean ONT_READ_LENGTH_MEAN (default=3000) \
                --ont-std ONT_READ_LENGTH_STD (default=2500) \
                --sr-platform ILLUMINA_PLATFORM (default='HS25') \
                --sr-mean ILLUMINA_INSERT_LENGTH_MEAN (default=400) \
                --sr-std ILLUMINA_INSERT_LENGTH_STD (default=125) \
                --sr-readlen ILLUMINA_READ_LENGTH (default=150)
```
##### The pipeline will generate simulated fastq file from both `Illumina` and `Oxford Nanopore Technology` platform, fasta file of both circular and linear DNA and the annotation files including the coordinate and coverage of each DNA.

### Tutorial
* [1. Notebook for compare statistics between given data and simulated data]((scripts/001_Simulated_Visualization.ipynb))
* [2. Shell scripts to run pipelines for eccDNA identification](scripts/000_Upstream_analysis_pipeline_detecting_eccDNA.ipynb)
* [3. Notebook for evaluation of performance of pipelines for short-read data analysis](scripts/002_ShortRead_Evaluation.ipynb)
* [4. Notebook for evaluation of performance of pipelines for long-read data analysis](scripts/003_LongRead_Evaluation.ipynb)


