{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4edbf009-0e9c-49f5-b053-d9183991ddcb",
   "metadata": {},
   "source": [
    "#### 实验名称：鉴定eccDNA的所有上游分析pipeline整理\n",
    "#### 实验目的：将benchmark工作所用的所有上游分析pipeline进行整理（NGS pipeline和TGS pipeline分别以NGS_100X和Nanopore_100X数据为例）\n",
    "#### 实验时间：2023-07-03\n",
    "#### 实验人员：罗淞文"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd586b4f-a046-4829-92af-6e0cbcc74adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NGS pipeline整理\n",
    "\n",
    "## Circle-Map pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J circle_map_test_NGS_100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/circle_map_test_NGS_100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/circle_map_test_NGS_100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/liuk0617/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate Circle-Map\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "R1fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R1.fq\n",
    "R2fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R2.fq\n",
    "reference_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38\n",
    "bwa_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/bwa/hg38\n",
    "result_dir=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/result/circle_map_result/NGS100X_01\n",
    "mkdir $result_dir\n",
    "cd $result_dir\n",
    "## Alignment of the reads to the reference genome\n",
    "bwa mem -t 40 -q $bwa_dir/hg38.fa $R1fastq $R2fastq > simple_ecc.sam\n",
    "## Preparing the files for Circle-Map\n",
    "samtools view -h simple_ecc.sam | sed -e '/@HD\\tVN:1.5\\tSO:unsorted\\tGO:query/d' > simple_ecc.sam\n",
    "samtools sort -@ 40 -n -o qname_simple_ecc.bam simple_ecc.sam\n",
    "samtools sort -@ 40 -o sorted_simple_ecc.bam simple_ecc.sam\n",
    "Circle-Map ReadExtractor -i qname_simple_ecc.bam -o circular_read_candidates.bam\n",
    "samtools sort -@ 40 -o sort_circular_read_candidates.bam circular_read_candidates.bam\n",
    "## Indexing the BAM files\n",
    "samtools index sort_circular_read_candidates.bam\n",
    "samtools index sorted_simple_ecc.bam\n",
    "## Detecting the circular DNA\n",
    "Circle-Map Realign -t 40 -i sort_circular_read_candidates.bam -qbam qname_simple_ecc.bam -sbam sorted_simple_ecc.bam -fasta $reference_dir/hg38.fa -o circle_NGS100X_01.bed\n",
    "echo End time: `date`\n",
    "\n",
    "## Circle_finder pipeline 整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J Circle_Finder_NGS100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/Circle_Finder_NGS100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/Circle_Finder_NGS100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate eccDNA\n",
    "bash_dir=/home/lifesci/luosongwen/eccDNA/Circle_finder-master\n",
    "cd /gpfs/home/lifesci/luosongwen/eccDNA/Circle_finder-master/Circle_DNA_NGS100X\n",
    "echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "bash $bash_dir/circle_finder-pipeline-bwa-mem-samblaster.sh 40 /home/lifesci/luosongwen/eccDNA/Circle_finder-master/hg38.fa /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R1.fq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R2.fq 10 Circle_final_NGS100X hg38\n",
    "echo Time is `date`\n",
    "\n",
    "## ecc_finder_NGS_map pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J ecc_finder_test_NGS100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_NGS100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_NGS100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate ecc_finder\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "R1fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R1.fq\n",
    "R2fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R2.fq\n",
    "Reference_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/bwa/hg38\n",
    "result_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/result/ecc_finder\n",
    "ecc_finder_path=/home/lifesci/luosongwen/eccDNA/ecc_finder\n",
    "cd $ecc_finder_path\n",
    "\n",
    "python ecc_finder.py map-sr $Reference_dir/hg38.fa $R1fastq $R2fastq -r $Reference_dir/hg38.fa -t 40  -o NGS100X_01_final_test_map\n",
    "echo End time: `date`\n",
    "\n",
    "## ecc_finder_NGS_asm pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J ecc_finder_test_NGS100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_NGS100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_NGS100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate ecc_finder\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "R1fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R1.fq\n",
    "R2fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R2.fq\n",
    "Reference_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/bwa/hg38\n",
    "result_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/result/ecc_finder\n",
    "ecc_finder_path=/home/lifesci/luosongwen/eccDNA/ecc_finder\n",
    "cd $ecc_finder_path\n",
    "\n",
    "python ecc_finder.py asm-sr $R1fastq $R2fastq -t 40 -o NGS100X_final_asm\n",
    "echo End time: `date`\n",
    "\n",
    "## ECCsplorer pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J eccsplorer_test_NGS100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/eccsplorer_test_NGS100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/eccsplorer_test_NGS100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "cd /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate eccsplorer\n",
    "python /home/lifesci/luosongwen/miniconda3/envs/eccsplorer/bin/ECCsplorer/ECCsplorer.py -out result/eccsplorer_result/final_NGS100X_01 /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R1.fq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/2nd_data/ratio/NGS100X_01_R2.fq -ref /home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38/hg38.clean.fa -log -m map\n",
    "echo End time is `date`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "935fd108-e597-406a-948e-499d45013445",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TGS pipeline整理\n",
    "\n",
    "## CReSIL pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J Cresil_nanopore100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/cresil_nanopore100X_01_test_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/cresil_nanopore100X_01_test_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate cresil\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "##Cresil_eccDNA_test\n",
    "## Go to an example folder\n",
    "cd /home/lifesci/luosongwen/eccDNA/TGS_tools/cresil_v2.0/example\n",
    "## Run trim \n",
    "cresil trim -t 40 -fq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq -r reference.mmi -o cresil_Nanopore100X_01_result\n",
    "## Run eccDNA identification for enriched data\n",
    "cresil identify -t 40 -fa reference.fa -fai reference.fa.fai -fq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq -trim cresil_Nanopore100X_01_result/trim.txt\n",
    "#cresil identify_wgls -t 40 -r reference.mmi -fa reference.fa -fai reference.fa.fai -fq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/3rd_simulate_data/WGS/nanoporeWGS100x.fastq -trim cresil_nanoporeWGS100x_result/trim.txt\n",
    "## Run eccDNA annotation\n",
    "cresil annotate -t 40 -rp reference.rmsk.bed -cg reference.cpg.bed -gb reference.gene.bed -identify cresil_Nanopore100X_01_result/eccDNA_final.txt\n",
    "## Run visualize eccDNA (ec1)\n",
    "#cresil visualize -t 4 -c ec1 -identify cresil_result/eccDNA_final.txt\n",
    "## Run Circos (ec1)\n",
    "#cd cresil_result/for_Circos/ec1\n",
    "#circos -noparanoid -conf circos.conf\n",
    "echo End time: `date`\n",
    "\n",
    "## NanoCircle pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J Nanocircle_nanopore100X\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/nanocircle_nanopore100X_test_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/nanocircle_nanopore100X_test_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate cresil\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "cd /home/lifesci/luosongwen/eccDNA/TGS_tools/cresil_nanocircle_v1.0/NanoCircle\n",
    "minimap2 -t 40 -ax map-ont --secondary=no /home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38/hg38.clean.fa /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq | samtools sort -> /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.bam\n",
    "samtools index /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.bam\n",
    "bedtools genomecov -bg -ibam /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.bam | bedtools merge -d 1000 -i stdin | sort -V -k1,1 -k2,2n > /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01_1000_cov.bed\n",
    "python /home/lifesci/luosongwen/eccDNA/TGS_tools/cresil_nanocircle_v1.0/NanoCircle/NanoCircle_arg.py Circles -i /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01_1000_cov.bed -b /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.bam -o /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01\n",
    "python /home/lifesci/luosongwen/eccDNA/TGS_tools/cresil_nanocircle_v1.0/NanoCircle/NanoCircle_arg.py Merge -i /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01_Complex.bed -o /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01_complex_merged.bed\n",
    "echo End time: `date`\n",
    "\n",
    "\n",
    "## ecc_finder_TGS_map pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J ecc_finder_test_map_nanopore100X\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_map_nanopore100X_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_map_nanopore100X_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate ecc_finder\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "R1fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq\n",
    "Reference_dir=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38\n",
    "result_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/result/ecc_finder\n",
    "ecc_finder_path=/home/lifesci/luosongwen/eccDNA/ecc_finder\n",
    "cd $ecc_finder_path\n",
    "python ecc_finder.py map-ont $Reference_dir/hg38.clean.fa $R1fastq -r $Reference_dir/hg38.clean.fa -t 40 -o Nanopore100X_final_map\n",
    "echo End time: `date`\n",
    "\n",
    "## ecc_finder_TGS_asm pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J ecc_finder_test_nanopore100X\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_nanopore100X_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/ecc_finder_test_nanopore100X_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate ecc_finder\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "R1fastq=/gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq\n",
    "Reference_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38\n",
    "result_dir=/home/lifesci/luosongwen/eccDNA/stimulate_data/result/ecc_finder\n",
    "ecc_finder_path=/home/lifesci/luosongwen/eccDNA/ecc_finder\n",
    "cd $ecc_finder_path\n",
    "python ecc_finder.py asm-ont $R1fastq -t 40 -o Nanopore100X_final_asm\n",
    "echo End time: `date`\n",
    "\n",
    "## eccDNA_RCA_nanopore pipeline整理\n",
    "#!/bin/sh\n",
    "#An example for serial job.\n",
    "#SBATCH -J eccDNA_RCA_nanopore100X_01\n",
    "#SBATCH -o /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/eccDNA_RCA_nanopore100X_01_%j.log\n",
    "#SBATCH -e /home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/logs/eccDNA_RCA_nanopore100X_01_%j.err\n",
    "#SBATCH --qos=aepmemqos\n",
    "#SBATCH --exclusive\n",
    "#SBATCH --time=240:00:00\n",
    "#SBATCH -N 1 -n 40 -p 2TB-AEP-Mem\n",
    "echo Running on $SBATCH_PARTITION paratation\n",
    "echo Start time is `date`\n",
    "source /home/lifesci/luosongwen/miniconda3/etc/profile.d/conda.sh\n",
    "conda activate cresil\n",
    "#echo Directory is $PWD\n",
    "echo This job runs on the following nodes:\n",
    "echo $SLURM_JOB_NODELIST\n",
    "echo This job has allocated $SLURM_JOB_CPUS_PER_NODE cpu core\n",
    "\n",
    "cd /home/lifesci/luosongwen/eccDNA/TGS_tools/eccDNA_RCA_nanopore\n",
    "minimap2 -t 40 -c /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38/hg38.clean.fa /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq > /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.paf\n",
    "./eccDNA_RCA_nanopore.py --fastq /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.fastq --paf /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/test_data/nanopore_01_data/NanoporeCircle100X_01.num.paf --info nanopore100X_01_num_info.tsv --seq nanopore100X_01_num.fa --var nanopore100X_01_num_var.tsv --reference /gpfs/home/lifesci/luosongwen/eccDNA/stimulate_data/Reference/UCSC/hg38/hg38.clean.fa --verbose | tee out.log\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e52c50f-7080-4072-a9d6-20eb6cd29f3b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
